File: starterAIs/starter.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Cell {
public:
    Cell () {
        neighbors.resize(6);
    }
    Cell (int cell_index, int richness, int neighbors) :
    cell_index{cell_index}, richness{richness}, neighbors{neighbors} {}

    void input() {
        cin >> cell_index >> richness;
        for (auto& n: neighbors) {
            cin >> n;
        }
    }
    int cell_index;
    int richness;
    vector<int> neighbors;
};

class Tree {
public:
    Tree () = default;
    Tree (int cell_index, int size, int is_mine, int is_dormant) :
        cell_index{cell_index}, size{size}, is_mine{is_mine}, is_dormant{is_dormant} {}
    void input() {
        int cellIndex;
        int size;
        bool isMine;
        bool isDormant;
        cin >> cellIndex >> size >> isMine >> isDormant;
    }
    int cell_index;
    int size;
    int is_mine;
    int is_dormant;
};

class Game {
private:
        int day = 0;
        int nutrients = 0;
        vector<Cell> board;
        vector<Tree> trees;
        vector<pair<string,int>> possible_actions;
        int mySun;
        int oppSun;
        int score;
        int oppScore;
        int oppIsWaiting;

public:
    void inputInitData() {
        int numberOfCells;
        cin >> numberOfCells;
        for (int i = 0; i < numberOfCells; i++) {
            Cell cell;
            cell.input();
            board.push_back(cell);
        }
    }

    void inputInfo() {
        // input game info
        cin >> day;
        cin >> nutrients;
        cin >> mySun >> score;
        cin >> oppSun >> oppScore >> oppIsWaiting;

        // input trees info
        trees.clear();
        int numberOfTrees;
        cin >> numberOfTrees;
        for (int i = 0; i < numberOfTrees; i++) {
            Tree tree;
            tree.input();
            trees.push_back(tree);
        }

        // input possible actions
        possible_actions.clear();
        int numberOfPossibleMoves;
        cin >> numberOfPossibleMoves;
        for (int i = 0; i < numberOfPossibleMoves; i++) {
            string type;
            int index = 0;
            cin >> type;
            if (type == "COMPLETE") {
                cin >> index;
                possible_actions.emplace_back(type, index);
            }
        }
    }
    //TODO: Please implement the algorithm in this function
    string compute_next_action() {
        string action = "WAIT"; // default
      
        // your algorithm !!

        return action;
    }
};

int main()
{
    Game game;
    game.inputInitData();

    while (true) {
        game.inputInfo();
        
        cout << game.compute_next_action() << endl;
    }
}

File: starterAIs/starter.pl
use strict;
use warnings;
use 5.32.1;
select STDOUT; $|=1;
my $game = new Game();
loadMap();
while (1) {
    loadTurn();
    say $game->getNextAction();
}

######################## subs

sub loadMap {
    chomp(my $number_of_cells = <STDIN>); # 37
    
    for my $i (0..$number_of_cells-1) {
        chomp(my $tokens=<STDIN>);
        my @parms = split / /,$tokens;
        my $cell = new Cell(shift @parms, shift @parms, \@parms);
        push @{ $game->{cells} }, $cell;
    }
}

sub loadTurn {
    chomp($game->{day} = <STDIN>);
    chomp($game->{nutrients} = <STDIN>);

    chomp(my $t = <STDIN>);
    ($game->{mySun}, $game->{myScore}) = split / /, $t;

    chomp($t = <STDIN>);
    ($game->{oppSun}, $game->{oppScore}, $game->{oppIsWaiting}) = split / /, $t;

    chomp(my $number_of_trees = <>);
    my @trees;
    for my $i (0..$number_of_trees-1) {
        my $tree = new Tree();
        chomp($t = <>);
        ($tree->{cellIndex}, $tree->{size}, $tree->{isMine}, $tree->{isDormant}) = split / /, $t;
        push @trees, $tree;
    }
    $game->{trees} = \@trees;
    
    chomp(my $number_of_actions = <>);
    my @actions;
    for my $i (0..$number_of_actions-1) {
        chomp($t = <>);
        my $action = Action->parse($t);
        push @actions, $action;
        #say STDERR $action->getAction();
    }
    $game->{possActions} = \@actions;
}

######################## classes

package Game;

sub new {
    my $type = shift;
    my (@cellArray, @actionArray, @treeArray);
    my $self = {
        day => 0,
        nutrients => 0,
        cells => \@cellArray,
        possActions => \@actionArray,
        trees => \@treeArray,
        mySun => 0,
        myScore => 0,
        oppSun => 0,
        oppScore => 0,
        oppIsWaiting => 0
    };
    return bless $self, $type;
}

sub getNextAction {
    my $self = shift;

    #Your logic to determine the next action goes here.

    return $self->{possActions}[0]->getAction();
}

package Cell;

sub new {
    my $type = shift;
    my $self = {
        index => shift,              
        richness => shift,
        neighbors => shift
    };
    return bless $self, $type;
}

package Tree;

sub new {
    my $type = shift;
    my $self = {
        cellIndex => shift,
        size => shift,
        isMine => shift,
        isDormant => shift
    };
    return bless $self, $type;
}

package Action;

use constant COMPLETE => "COMPLETE";
use constant GROW => "GROW";
use constant SEED => "SEED";
use constant WAIT => "WAIT";

sub new {
    my $type = shift;
    my $self = {
        type => shift,
        target => shift,
        source => shift
    };
    return bless $self, $type;
}

sub parse {
    my ($self, $line) = @_;
    my @parts = split / /, $line;

    if ($parts[0] eq WAIT) {
        return new Action(WAIT);
    }

    if ($parts[0] eq SEED) {
        return new Action(SEED, $parts[2], $parts[1]);
    }

    return new Action($parts[0], $parts[1]);
}

sub getAction {
    my ($self) = @_;

    if ($self->{type} eq WAIT) {
        return WAIT;
    }

    if ($self->{type} eq SEED) {
        return SEED." ".$self->{target}." ".$self->{source};
    }

    return $self->{type}." ".$self->{target};   
}


